[MRG] Add pprint for estimators - continued

NicolasHug opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#pull-request-checklist
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->

This is the continuation of #9099 that aims to have better printing representations for estimators

Close #9099
Close #9039
Close #7618
Fix #6323

#### What does this implement/fix? Explain your changes.

- ~~Fixed some line length issues~~
- ~~Factored code for Estimator and Pipeline into a single method~~
- Completely refactored code to use extend the built-in `PrettyPrinter` class
- Added some tests
- Added some doc

#### Any other comments?

I marked this as WIP but a first review may be useful. I tried to make the code easier to understand but it's always a bit hard with those kind of functionalities.

Line length limit is still not perfect, see e.g. the test with `RFE(...)` that goes beyond 80 chars because the printed estimator is actually an argument.

#### Questions

- Should the file `_pprint.py` be in `utils` instead of at the root? (test file would follow)
- Should I remove the `_pprint()` method in `base.py` and use the Formatter class in `__repr__` instead?

amueller committed:
add pprint for estimators

amueller committed:
strip color from length, add color option

NicolasHug committed:
Minor cleaning, fixes, factoring and docs

NicolasHug committed:
Added some basic tests

NicolasHug committed:
Fixed line length issue

NicolasHug committed:
fixed flake8 and added visual test for review

NicolasHug committed:
Fixed test

NicolasHug committed:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn into pr/9099

NicolasHug committed:
Fixed Python 2 issues (inspect.signature import)

NicolasHug committed:
Trying to fix flake8 again

jnothman commented:
best if you ping after the 0.20 RC is out

>

amueller commented:
can you add a test with a pipeline inside a grid-search please? And the pipeline maybe has a ColumnTransformer? ;)

NicolasHug committed:
Added special repr for functions

NicolasHug committed:
Added some other visual tests

NicolasHug committed:
Changed _format_function in to _format_callable

because callable() returns True also for class objects (which we want to
reprensent with their name as well anyway)

NicolasHug committed:
Consistent output in Python 2 and 3

jnothman commented:
Don't worry about python 2 support. It will be dropped in the next release.

NicolasHug committed:
WIP

NicolasHug committed:
Now using the builtin pprint module

NicolasHug committed:
pep8

NicolasHug committed:
Added changed_only param

NicolasHug commented:
So due to some limitations of the current method, I completely changed the implementation and relied on pprint.PrettyPrinter instead.

Here are some output examples:

<details>

```
LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, max_iter=100, multi_class='ovr',
                   n_jobs=1, penalty='l2', random_state=None,
                   solver='liblinear', tol=0.0001, verbose=0, warm_start=False)

Pipeline(memory=None,
         steps=['reduce_dim': PCA(copy=True, iterated_power='auto',
                                  n_components=None, random_state=None,
                                  svd_solver='auto', tol=0.0, whiten=False),
                'classify': LinearSVC(C=1.0, class_weight=None, dual=True,
                                      fit_intercept=True, intercept_scaling=1,
                                      loss='squared_hinge', max_iter=1000,
                                      multi_class='ovr', penalty='l2',
                                      random_state=None, tol=0.0001,
                                      verbose=0)])

RFE(estimator=RFE(estimator=RFE(estimator=RFE(estimator=RFE(estimator=RFE(estimator=RFE(estimator=LogisticRegression(C=1.0,
                                                                                                                     class_weight=None,
                                                                                                                     dual=False,
                                                                                                                     fit_intercept=True,
                                                                                                                     intercept_scaling=1,
                                                                                                                     max_iter=100,
                                                                                                                     multi_class='ovr',
                                                                                                                     n_jobs=1,
                                                                                                                     penalty='l2',
                                                                                                                     random_state=None,
                                                                                                                     solver='liblinear',
                                                                                                                     tol=0.0001,
                                                                                                                     verbose=0,
                                                                                                                     warm_start=False),
                                                                                        n_features_to_select=None,
                                                                                        step=1,
                                                                                        verbose=0),
                                                                          n_features_to_select=None,
                                                                          step=1,
                                                                          verbose=0),
                                                            n_features_to_select=None,
                                                            step=1, verbose=0),
                                              n_features_to_select=None, step=1,
                                              verbose=0),
                                n_features_to_select=None, step=1, verbose=0),
                  n_features_to_select=None, step=1, verbose=0),
    n_features_to_select=None, step=1, verbose=0)

GridSearchCV(cv=5, error_score='raise-deprecating',
             estimator=SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
                           decision_function_shape='ovr', degree=3,
                           gamma='auto_deprecated', kernel='rbf', max_iter=-1,
                           probability=False, random_state=None, shrinking=True,
                           tol=0.001, verbose=False),
             fit_params=None, iid='warn', n_jobs=1,
             param_grid=[{'C': [1, 10, 100, 1000], 'gamma': [0.001, 0.0001],
                          'kernel': ['rbf']},
                         {'C': [1, 10, 100, 1000], 'kernel': ['linear']}],
             pre_dispatch='2*n_jobs', refit=True, return_train_score='warn',
             scoring=None, verbose=0)

GridSearchCV(cv=3, error_score='raise-deprecating',
             estimator=Pipeline(memory=None,
                                steps=['reduce_dim': PCA(copy=True,
                                                         iterated_power='auto',
                                                         n_components=None,
                                                         random_state=None,
                                                         svd_solver='auto',
                                                         tol=0.0,
                                                         whiten=False),
                                       'classify': LinearSVC(C=1.0,
                                                             class_weight=None,
                                                             dual=True,
                                                             fit_intercept=True,
                                                             intercept_scaling=1,
                                                             loss='squared_hinge',
                                                             max_iter=1000,
                                                             multi_class='ovr',
                                                             penalty='l2',
                                                             random_state=None,
                                                             tol=0.0001,
                                                             verbose=0)]),
             fit_params=None, iid='warn', n_jobs=1,
             param_grid=[{'classify__C': [1, 10, 100, 1000],
                          'reduce_dim': [PCA(copy=True, iterated_power=7,
                                             n_components=None,
                                             random_state=None,
                                             svd_solver='auto', tol=0.0,
                                             whiten=False),
                                         NMF(alpha=0.0, beta_loss='frobenius',
                                             init=None, l1_ratio=0.0,
                                             max_iter=200, n_components=None,
                                             random_state=None, shuffle=False,
                                             solver='cd', tol=0.0001,
                                             verbose=0)],
                          'reduce_dim__n_components': [2, 4, 8]},
                         {'classify__C': [1, 10, 100, 1000],
                          'reduce_dim': [SelectKBest(k=10,
                                                     score_func=<function chi2 at 0x7f77d68c18c8>)],
                          'reduce_dim__k': [2, 4, 8]}],
             pre_dispatch='2*n_jobs', refit=True, return_train_score='warn',
             scoring=None, verbose=0)
```
</details>

--------

The default is to indent after the name of the estimator, but it's expensive in line width. Unfortunately, `PrettyPrinter` does not really support other forms of indentations:

>  The amount of indentation added for each recursive level is specified by indent; the default is one. Other values can cause output to look a little odd

From the [docs](https://docs.python.org/3/library/pprint.html). Example of discrepancy:

```
pp = PrettyPrinter(indent=4, compact=True) 
pp.pprint(list(range(30))) 
```

will give

```
[   0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20,
    21, 22, 23, 24, 25, 26, 27, 28, 29]
```

Here is what the output looks like on the same examples when indenting with 4 characters:

<details>

```
LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
    intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
    penalty='l2', random_state=None, solver='liblinear', tol=0.0001, verbose=0,
    warm_start=False)

Pipeline(memory=None,
    steps=[   'reduce_dim': PCA(copy=True, iterated_power='auto',
                                n_components=None, random_state=None,
                                svd_solver='auto', tol=0.0, whiten=False),
              'classify': LinearSVC(C=1.0, class_weight=None, dual=True,
                              fit_intercept=True, intercept_scaling=1,
                              loss='squared_hinge', max_iter=1000,
                              multi_class='ovr', penalty='l2',
                              random_state=None, tol=0.0001, verbose=0)])

RFE(estimator=RFE(estimator=RFE(estimator=RFE(estimator=RFE(estimator=RFE(estimator=RFE(estimator=LogisticRegression(C=1.0,
                                                                                                      class_weight=None,
                                                                                                      dual=False,
                                                                                                      fit_intercept=True,
                                                                                                      intercept_scaling=1,
                                                                                                      max_iter=100,
                                                                                                      multi_class='ovr',
                                                                                                      n_jobs=1,
                                                                                                      penalty='l2',
                                                                                                      random_state=None,
                                                                                                      solver='liblinear',
                                                                                                      tol=0.0001,
                                                                                                      verbose=0,
                                                                                                      warm_start=False),
                                                                                        n_features_to_select=None,
                                                                                        step=1,
                                                                                        verbose=0),
                                                                          n_features_to_select=None,
                                                                          step=1,
                                                                          verbose=0),
                                                            n_features_to_select=None,
                                                            step=1, verbose=0),
                                              n_features_to_select=None, step=1,
                                              verbose=0),
                                n_features_to_select=None, step=1, verbose=0),
                  n_features_to_select=None, step=1, verbose=0),
    n_features_to_select=None, step=1, verbose=0)

GridSearchCV(cv=5, error_score='raise-deprecating',
    estimator=SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
                  decision_function_shape='ovr', degree=3,
                  gamma='auto_deprecated', kernel='rbf', max_iter=-1,
                  probability=False, random_state=None, shrinking=True,
                  tol=0.001, verbose=False),
    fit_params=None, iid='warn', n_jobs=1,
    param_grid=[   {   'C': [1, 10, 100, 1000], 'gamma': [0.001, 0.0001],
                       'kernel': ['rbf']},
                   {'C': [1, 10, 100, 1000], 'kernel': ['linear']}],
    pre_dispatch='2*n_jobs', refit=True, return_train_score='warn',
    scoring=None, verbose=0)

GridSearchCV(cv=3, error_score='raise-deprecating',
    estimator=Pipeline(memory=None,
                  steps=[   'reduce_dim': PCA(copy=True, iterated_power='auto',
                                              n_components=None,
                                              random_state=None,
                                              svd_solver='auto', tol=0.0,
                                              whiten=False),
                            'classify': LinearSVC(C=1.0, class_weight=None, dual=True,
                                            fit_intercept=True,
                                            intercept_scaling=1,
                                            loss='squared_hinge', max_iter=1000,
                                            multi_class='ovr', penalty='l2',
                                            random_state=None, tol=0.0001,
                                            verbose=0)]),
    fit_params=None, iid='warn', n_jobs=1,
    param_grid=[   {   'classify__C': [1, 10, 100, 1000],
                       'reduce_dim': [   PCA(copy=True, iterated_power=7,
                                             n_components=None,
                                             random_state=None,
                                             svd_solver='auto', tol=0.0,
                                             whiten=False),
                                         NMF(alpha=0.0, beta_loss='frobenius',
                                             init=None, l1_ratio=0.0,
                                             max_iter=200, n_components=None,
                                             random_state=None, shuffle=False,
                                             solver='cd', tol=0.0001,
                                             verbose=0)],
                       'reduce_dim__n_components': [2, 4, 8]},
                   {   'classify__C': [1, 10, 100, 1000],
                       'reduce_dim': [   SelectKBest(k=10,
                                             score_func=<function chi2 at 0x7f4a124bd8c8>)],
                       'reduce_dim__k': [2, 4, 8]}],
    pre_dispatch='2*n_jobs', refit=True, return_train_score='warn',
    scoring=None, verbose=0)
```
</details>

I personally like the first output. It might render quite long strings when estimators are nested but I don't think that it's a major issue, and it's very easy to read.

jnothman commented:
Yes, I much prefer the first. Good idea to reuse PrettyPrinter. Will that be configurable for us wanting to hide default parameter settings, etc?

NicolasHug commented:
Everything is encapsulated in `BaseEstimator.__repr__()` so we can hide the default parameters, if that what you mean?

I noticed that the `steps` is formatted as a list of `key: value`s instead of a list of tuples. Is it OK like this or do we want to change it? I guess it's slightly clearer this way but it prevents the string from being evaluated with `eval` for example.

jnothman commented:
I think it should be a list of tuples and am surprised that it is not.

NicolasHug committed:
Fixed printing when string would fit in less than line width

NicolasHug committed:
Fixed printing of steps parameter

NicolasHug committed:
Fixed changed_only param for short estimators

NicolasHug committed:
fixed pep8

NicolasHug commented:
I fixed it, here is the new look:

<details>

```
LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, max_iter=100, multi_class='ovr',
                   n_jobs=1, penalty='l2', random_state=None,
                   solver='liblinear', tol=0.0001, verbose=0, warm_start=False)

Pipeline(memory=None,
         steps=[('reduce_dim',
                 PCA(copy=True, iterated_power='auto', n_components=None,
                     random_state=None, svd_solver='auto', tol=0.0,
                     whiten=False)),
                ('classify',
                 LinearSVC(C=1.0, class_weight=None, dual=True,
                           fit_intercept=True, intercept_scaling=1,
                           loss='squared_hinge', max_iter=1000,
                           multi_class='ovr', penalty='l2', random_state=None,
                           tol=0.0001, verbose=0))])

RFE(estimator=RFE(estimator=RFE(estimator=RFE(estimator=RFE(estimator=RFE(estimator=RFE(estimator=LogisticRegression(C=1.0,
                                                                                                                     class_weight=None,
                                                                                                                     dual=False,
                                                                                                                     fit_intercept=True,
                                                                                                                     intercept_scaling=1,
                                                                                                                     max_iter=100,
                                                                                                                     multi_class='ovr',
                                                                                                                     n_jobs=1,
                                                                                                                     penalty='l2',
                                                                                                                     random_state=None,
                                                                                                                     solver='liblinear',
                                                                                                                     tol=0.0001,
                                                                                                                     verbose=0,
                                                                                                                     warm_start=False),
                                                                                        n_features_to_select=None,
                                                                                        step=1,
                                                                                        verbose=0),
                                                                          n_features_to_select=None,
                                                                          step=1,
                                                                          verbose=0),
                                                            n_features_to_select=None,
                                                            step=1, verbose=0),
                                              n_features_to_select=None, step=1,
                                              verbose=0),
                                n_features_to_select=None, step=1, verbose=0),
                  n_features_to_select=None, step=1, verbose=0),
    n_features_to_select=None, step=1, verbose=0)

GridSearchCV(cv=5, error_score='raise-deprecating',
             estimator=SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
                           decision_function_shape='ovr', degree=3,
                           gamma='auto_deprecated', kernel='rbf', max_iter=-1,
                           probability=False, random_state=None, shrinking=True,
                           tol=0.001, verbose=False),
             fit_params=None, iid='warn', n_jobs=1,
             param_grid=[{'C': [1, 10, 100, 1000], 'gamma': [0.001, 0.0001],
                          'kernel': ['rbf']},
                         {'C': [1, 10, 100, 1000], 'kernel': ['linear']}],
             pre_dispatch='2*n_jobs', refit=True, return_train_score='warn',
             scoring=None, verbose=0)

GridSearchCV(cv=3, error_score='raise-deprecating',
             estimator=Pipeline(memory=None,
                                steps=[('reduce_dim',
                                        PCA(copy=True, iterated_power='auto',
                                            n_components=None,
                                            random_state=None,
                                            svd_solver='auto', tol=0.0,
                                            whiten=False)),
                                       ('classify',
                                        LinearSVC(C=1.0, class_weight=None,
                                                  dual=True, fit_intercept=True,
                                                  intercept_scaling=1,
                                                  loss='squared_hinge',
                                                  max_iter=1000,
                                                  multi_class='ovr',
                                                  penalty='l2',
                                                  random_state=None, tol=0.0001,
                                                  verbose=0))]),
             fit_params=None, iid='warn', n_jobs=1,
             param_grid=[{'classify__C': [1, 10, 100, 1000],
                          'reduce_dim': [PCA(copy=True, iterated_power=7,
                                             n_components=None,
                                             random_state=None,
                                             svd_solver='auto', tol=0.0,
                                             whiten=False),
                                         NMF(alpha=0.0, beta_loss='frobenius',
                                             init=None, l1_ratio=0.0,
                                             max_iter=200, n_components=None,
                                             random_state=None, shuffle=False,
                                             solver='cd', tol=0.0001,
                                             verbose=0)],
                          'reduce_dim__n_components': [2, 4, 8]},
                         {'classify__C': [1, 10, 100, 1000],
                          'reduce_dim': [SelectKBest(k=10,
                                                     score_func=<function chi2 at 0x7f046321a950>)],
                          'reduce_dim__k': [2, 4, 8]}],
             pre_dispatch='2*n_jobs', refit=True, return_train_score='warn',
             scoring=None, verbose=0)
```
</details>


I'll put this as MRG for reviews, and if it gets approved I'll start updating the docs with the new formatting to fix the tests.

jnothman commented:
I think that looks good...

jnothman commented:
Does it need a code review at this point?

NicolasHug commented:
Yes, if the output is OK for everyone then the current code can be reviewed.

Note that `_safe_repr` and `_format_params_or_dict_items` are almost entirely copy pasted from the builtin `pprint` modulem so even tough they may be a bit opaque it probably doesn't make much sense to change them.

amueller commented:
Test failures on python3 are from whitespace changes, which probably require ``#doctest +IGNORE_WHITESPACE``. Python2.7 has syntax issues, but I guess we don't really care about that at this point?

amueller commented:
I think this is a giant improvement! we should do this for now, ideally with option to hide the defaults.

amueller commented:
I think it would be good to say explicitly in the class what we changed and why.
Basically this is allowing the ``compact`` parameter to be applied to dictionaries, IIRC.

amueller commented:
oh and I would like the behavior re changed params to be controlled via a ``set_config`` parameter.

NicolasHug committed:
Added some more description in docstring

NicolasHug committed:
changed_only is now an option from set_config()

NicolasHug committed:
Put _pprint.py into sklearn/utils, added tests

NicolasHug committed:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn into pr/9099

amueller commented:
let me know when you want another round of comments?

sklearn-lgtm commented:
This pull request **introduces 1 alert** when merging eb9a171d806fb1e9664fbb17d54c77069bdd4158 into a3616f65dc550b2c93888be478ee098df2b01888 - [view on LGTM.com](https://lgtm.com/projects/g/scikit-learn/scikit-learn/rev/pr-eb3c7c85959f3134bd4a57e1f57733dac3ff63f2)

**new alerts:**

* 1 for Syntax error

---

*Comment posted by [LGTM.com](https://lgtm.com)*

NicolasHug committed:
Added doctest NORMALIZE_WHITESPACE where needed

sklearn-lgtm commented:
This pull request **introduces 1 alert** when merging 8b5b283369c2d610e5bf3ed8afa794849ed34a4e into a3616f65dc550b2c93888be478ee098df2b01888 - [view on LGTM.com](https://lgtm.com/projects/g/scikit-learn/scikit-learn/rev/pr-1e9a02ed75c5f10fda5b23057bbb24f8a3a4a05d)

**new alerts:**

* 1 for Syntax error

---

*Comment posted by [LGTM.com](https://lgtm.com)*

NicolasHug committed:
Fixed tests

NicolasHug commented:
I think I finished fixing all the tests so you can go for another round of comments.

Those changes are clearly >= 3.5 only so tests will still fail for previous versions. This is due to the builtin pprint module that was reworked.

sklearn-lgtm commented:
This pull request **introduces 1 alert** when merging f0ed05fe930c94aaf453870f3034eac89596f460 into 94c70ff235a19312063ef089ef587957f40db656 - [view on LGTM.com](https://lgtm.com/projects/g/scikit-learn/scikit-learn/rev/pr-2e333820cf6650ff3df74282da522856acdaae6d)

**new alerts:**

* 1 for Syntax error

---

*Comment posted by [LGTM.com](https://lgtm.com)*

NicolasHug committed:
fix test-doc

sklearn-lgtm commented:
This pull request **introduces 1 alert** when merging e785d22b3f8646348377dca88f22d53df8344ca8 into 59b15c524914cbc73f347da1565e199634960f1f - [view on LGTM.com](https://lgtm.com/projects/g/scikit-learn/scikit-learn/rev/pr-44c19ccf8f97ff19633e94056a88a929e05a27df)

**new alerts:**

* 1 for Syntax error

---

*Comment posted by [LGTM.com](https://lgtm.com)*

NicolasHug committed:
fixing test that passed before....

sklearn-lgtm commented:
This pull request **introduces 1 alert** when merging b64258e13f434e04648f35ceed571459f8aaafc5 into 59b15c524914cbc73f347da1565e199634960f1f - [view on LGTM.com](https://lgtm.com/projects/g/scikit-learn/scikit-learn/rev/pr-1f5454614649da0279e2434e76d17e506ba30922)

**new alerts:**

* 1 for Syntax error

---

*Comment posted by [LGTM.com](https://lgtm.com)*

amueller reviewed (approved):
looks good, thanks! Not sure if we should merge now or wait for 0.20.1 when we'll disable old python on the CI.

amueller commented:
can you fix the conflicts please?

NicolasHug committed:
Merge branch 'master' into pr/9099

sklearn-lgtm commented:
This pull request **introduces 1 alert** when merging 14eac3b7cf7e32644d2c0086139ec733793b5f5e into 4e2da4af927567d94ec1874c817e7e948be5dda8 - [view on LGTM.com](https://lgtm.com/projects/g/scikit-learn/scikit-learn/rev/pr-15ea2e3b616c8cb52a3f7e8a7cb097a04541099a)

**new alerts:**

* 1 for Syntax error

---

*Comment posted by [LGTM.com](https://lgtm.com)*

NicolasHug committed:
Fixed tests

sklearn-lgtm commented:
This pull request **introduces 1 alert** when merging 2932be88843aee5144952c517e7e343a8dd01b23 into 4e2da4af927567d94ec1874c817e7e948be5dda8 - [view on LGTM.com](https://lgtm.com/projects/g/scikit-learn/scikit-learn/rev/pr-77e0980092211ed711626a16ff15280390351e8d)

**new alerts:**

* 1 for Syntax error

---

*Comment posted by [LGTM.com](https://lgtm.com)*

jnothman commented:
Let's release 0.20.1 first.

jnothman commented on sklearn/_config.py:
```diff
@@ -43,11 +45,21 @@ def set_config(assume_finite=None, working_memory=None):
 
         .. versionadded:: 0.20
 
+    print_changed_only : bool, optional
```
would this be clearer as print_defaults?

NicolasHug replied:
I don't think so because the meaning of the parameter is really "only show the parameters that don't have their default values, i.e. the ones that are changed"

"print_defaults" would mean something quite different.

Maybe "hide_defaults"? Or "print_hide_defaults"?

jnothman commented on sklearn/_config.py:
```diff
@@ -43,11 +45,21 @@ def set_config(assume_finite=None, working_memory=None):
+    print_changed_only : bool, optional
+        If True, only the parameters that were set to non-default
+        values will be printed when printing an estimator. For example,
+        ``print(SCV())`` while True will only print 'SVC()' while the default
```
*SCV -> SVC

jnothman commented on sklearn/tests/test_base.py:
```diff
@@ -178,7 +178,7 @@ def test_repr():
 
     some_est = T(a=["long_params"] * 1000)
-    assert_equal(len(repr(some_est)), 415)
+    assert_equal(len(repr(some_est)), 16015)
```
I think the point of this test was to check that the repr would not be excessively long, and that ellipsis would be used. I'm not convinced that this change is faithful to that concern

NicolasHug replied:
You're correct.
We're not using ellipsis anymore, should we add it?

I doubt it would be really useful now that the pprint is correctly formatted, supposing that we would almost never have extremely long reprs.

amueller replied:
We do by default, only with the short repr we wouldn't. Maybe we could switch to the short repr if the full repr is too long? That seems more useful than ellipsis?

jnothman replied:
What do you mean by the short repr? This test implies that the default repr is much longer than the current default

amueller replied:
by short I meant with ``changed_only=True``

jnothman replied:
What if it's a DictVectorizer with a provided vocabulary?

amueller replied:
I have no idea what the repr for this is in this pr. @NicolasHug ?
We might want to limit some things in length more than others maybe?

jnothman replied:
That is, we should bound use ellipsis for non-estimators

NicolasHug replied:
What do you mean a DictVectorizer with provided vocabulary?

`__repr__` (in this PR and in the current implem) only cares about the arguments passed to `__init__`, not those to `fit`, so the `vocabulary_` attribute would be ignored:

```python
from sklearn.feature_extraction import DictVectorizer
v = DictVectorizer(sparse=False)
D = [{'foo': 1, 'bar': 2}, {'foo': 3, 'baz': 1}]
v.fit(D)
v
```

```
DictVectorizer(dtype=<class 'numpy.float64'>, separator='=', sort=True,
               sparse=False)
```


Also what I meant by
>  We're not using ellipsis anymore

is that this PR won't do ellipsis. Of course the current implementation still does.

I like @amueller's idea: the default value of `print_changed_only` could be `'auto'`, which would use `True` iff the repr is more than x characters, else `False`.

amueller replied:
In CountVectorizer (but not DictVectorizer) you can pass a vocabulary in ``__init__``: https://scikit-learn.org/dev/modules/generated/sklearn.feature_extraction.text.CountVectorizer.html

NicolasHug replied:
```python
from sklearn.feature_extraction.text import CountVectorizer
vocabulary = {i: i for i in range(100)}
vectorizer = CountVectorizer(vocabulary=vocabulary)
vectorizer
```

```
CountVectorizer(analyzer='word', binary=False, decode_error='strict',
                dtype=<class 'numpy.int64'>, encoding='utf-8', input='content',
                lowercase=True, max_df=1.0, max_features=None, min_df=1,
                ngram_range=(1, 1), preprocessor=None, stop_words=None,
                strip_accents=None, token_pattern='(?u)\\b\\w\\w+\\b',
                tokenizer=None,
                vocabulary={0: 0, 1: 1, 2: 2, 3: 3, 4: 4, 5: 5, 6: 6, 7: 7,
                            8: 8, 9: 9, 10: 10, 11: 11, 12: 12, 13: 13, 14: 14,
                            15: 15, 16: 16, 17: 17, 18: 18, 19: 19, 20: 20,
                            21: 21, 22: 22, 23: 23, 24: 24, 25: 25, 26: 26,
                            27: 27, 28: 28, 29: 29, 30: 30, 31: 31, 32: 32,
                            33: 33, 34: 34, 35: 35, 36: 36, 37: 37, 38: 38,
                            39: 39, 40: 40, 41: 41, 42: 42, 43: 43, 44: 44,
                            45: 45, 46: 46, 47: 47, 48: 48, 49: 49, 50: 50,
                            51: 51, 52: 52, 53: 53, 54: 54, 55: 55, 56: 56,
                            57: 57, 58: 58, 59: 59, 60: 60, 61: 61, 62: 62,
                            63: 63, 64: 64, 65: 65, 66: 66, 67: 67, 68: 68,
                            69: 69, 70: 70, 71: 71, 72: 72, 73: 73, 74: 74,
                            75: 75, 76: 76, 77: 77, 78: 78, 79: 79, 80: 80,
                            81: 81, 82: 82, 83: 83, 84: 84, 85: 85, 86: 86,
                            87: 87, 88: 88, 89: 89, 90: 90, 91: 91, 92: 92,
                            93: 93, 94: 94, 95: 95, 96: 96, 97: 97, 98: 98,
                            99: 99})
```

The whole dict is printed even if we go up to `1000` instead of `100`. I'm afraid the `print_only_changed` parameter won't be of any help here :/

amueller replied:
Looks like we do need "...". Maybe only for non-estimators though?

jnothman replied:
And this current test is considering the case of a non-estimator parameter, so it should be more or less maintained, even if the number of characters differs.

NicolasHug replied:
Got it, I'll try to add ellipsis for non-estimators soon

NicolasHug replied:
Well... I looked into it and I think that adding ellipsis for non-estimator objects is going to be unreasonably difficult. Maybe even impossible.

The main reason is that, e.g. for a dict, `pprint` never returns a string representing this object. It will instead add '{' to a stream (typically stdout), then it will add all the key-values representations chunk by chunk, and then '}'. 

So there is no way to do something like

```
if len(my_dict_repr) > limit:
    # add ellipsis
```

because `my_dict_repr` does not exist.

We could easily add the ellipsis **after** the call to `pprint`, but would likely "cut" objects in arbitrary places.

amueller replied:
how about we call pprint to create the repr, then check if it's too long, and if it's too long we throw it out and create a new one using more and more aggressive heuristics that make it shorter.
The first one is ``changed_only=True``, the second one could be additionally replacing all non-scalar values by ellipsis and the fallback is Just print the estimator name with ellipsis ``CountVectorizer(...)```.

NicolasHug replied:
I think IPython uses its own [pprint](https://ipython.readthedocs.io/en/stable/api/generated/IPython.lib.pretty.html) module.

We could count the number of characters returned by `write` but that would require changing every single call to `write`, so we'd need to copy/paste the whole `pprint` module.

Even then, how do we alter the stream afterwards? As far as I can tell, going back in a `StringIO` object with `seek` is impossible:

```
seek(pos, whence=0, /) method of _io.StringIO instance
    Change stream position.
    
    Seek to character offset pos relative to position indicated by whence:
        0  Start of stream (the default).  pos should be >= 0;
        1  Current position - pos must be 0;
        2  End of stream - pos must be 0.
    Returns the new absolute position.
```

NicolasHug replied:
How about this:

The number of non-estimator parameters that are potentially very long (such as the `vocabulary` param of `CountVectorizer`) is pretty small, right?

So instead of trying to solve the problem in `PrettyPrinter`, we could try to solve it upstream, namely in a custom `get_param` function where we would replace those potentially long parameters by a shortened version. In the case of `vocabulary` that could be, for example, keeping only a subset of key-values, or replacing it by an empty dict. Or by a 'custom_vocabulary' str object.

It's really a case-by-case basis but I don't think we would have a lot of such cases?

This could be an additional step to @amueller's proposed process above:
- try pprint
- try again with those potentially long parameters removed
- try again with changed_only=True (and without the long params)
- ...

NicolasHug replied:
Here are a few examples with @amueller's idea of using ellipsis on non-basic types (lists, dicts, tuples, etc.)

It's pretty easy to implement, I can push changes if it's satisfactory for you. This fixes the `CountVectorizer/vocabulary` issue but `Pipeline` and `GridSearch` aren't great.

<details>

```
Pipeline(memory=None,
         steps=...)

RFE(estimator=RFE(estimator=RFE(estimator=RFE(estimator=RFE(estimator=RFE(estimator=RFE(estimator=LogisticRegression(C=1.0,
                                                                                                                     class_weight=None,
                                                                                                                     dual=False,
                                                                                                                     fit_intercept=True,
                                                                                                                     intercept_scaling=1,
                                                                                                                     max_iter=100,
                                                                                                                     multi_class='warn',
                                                                                                                     n_jobs=None,
                                                                                                                     penalty='l2',
                                                                                                                     random_state=None,
                                                                                                                     solver='warn',
                                                                                                                     tol=0.0001,
                                                                                                                     verbose=0,
                                                                                                                     warm_start=False),
                                                                                        n_features_to_select=None,
                                                                                        step=1,
                                                                                        verbose=0),
                                                                          n_features_to_select=None,
                                                                          step=1,
                                                                          verbose=0),
                                                            n_features_to_select=None,
                                                            step=1, verbose=0),
                                              n_features_to_select=None, step=1,
                                              verbose=0),
                                n_features_to_select=None, step=1, verbose=0),
                  n_features_to_select=None, step=1, verbose=0),
    n_features_to_select=None, step=1, verbose=0)


GridSearchCV(cv=5, error_score='raise-deprecating',
             estimator=SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
                           decision_function_shape='ovr', degree=3,
                           gamma='auto_deprecated', kernel='rbf', max_iter=-1,
                           probability=False, random_state=None, shrinking=True,
                           tol=0.001, verbose=False),
             iid='warn', n_jobs=None,
             param_grid=...,
             pre_dispatch='2*n_jobs', refit=True, return_train_score=False,
             scoring=None, verbose=0)


GridSearchCV(cv=3, error_score='raise-deprecating',
             ('estimator', Pipeline(memory=None,
         steps=...)),
             iid='warn', n_jobs=1,
             param_grid=...,
             pre_dispatch='2*n_jobs', refit=True, return_train_score=False,
             scoring=None, verbose=0)

CountVectorizer(analyzer='word', binary=False, decode_error='strict',
                dtype=<class 'numpy.int64'>, encoding='utf-8', input='content',
                lowercase=True, max_df=1.0, max_features=None, min_df=1,
                ngram_range=(1, 1), preprocessor=None, stop_words=None,
                strip_accents=None, token_pattern='(?u)\\b\\w\\w+\\b',
                tokenizer=None,
                vocabulary=...)

```
</details>

amueller replied:
I would always try the changed_only first and if that's not short enough, I'd do changed_only plus this heuristic. So these would look quite different, I think?

NicolasHug replied:
They would look slightly shorter yes. But we would still be using ellipsis on the `steps` or `param_grid` parameters which IMO makes the representations not very informative.

amueller replied:
yeah but if they are super long otherwise they won't be informative anyway, right?

NicolasHug replied:
Well in the case of `CountVectorizer/vocabulary` for example, what may make the `__repr__` super long is likely to be  `vocabulary`, not the `steps` or `param_grid` parameters.

Ideally, we should use ellipsis on `vocabulary` only. If it's still too long they we could use ellipsis more. But that would require hardcoding which parameters should be "ellipsed" first.

jnothman replied:
I don't think we need to hardcode which parameters should be ellipsed. If the vocabulary has more than `MAX_ELEMENTS_TO_SHOW` entries then we show a few and elide the rest.

NicolasHug replied:
@amueller , @jnothman ,

As of 5e23560 I have added:

- Ellipsis based on the number of elements in lists, tuple, dicts
- `__repr__` now tries in this order:
  - the original representation
  - repr with ellipsis on sequences with a lot of elements
  - repr with ellipsis on sequences with a lot of elements + changed_only = True
  - repr with ellipsis on sequences with a lot of elements + changed_only = True + bruteforce ellipsis that cuts the repr right in the middle.
- tests for all these features

Hope this will be it ^^ 🤞🤞🤞🤞

NicolasHug replied:
I don't think it adds a lot of extra complexity, it's just a few lines in `__repr__`. We've had the `changed_only` logic right from the beginning of this PR, now we just force it if the repr is too long.

amueller replied:
I think it's a nice place to show-case that functionality and it doesn't look that complex to me tbh. You mean complex in terms of code or complex for the user to understand?

jnothman replied:
I mean unnecessary and unjustified and hence complex by all measures. But I should check the code to make sure my understanding is correct

jnothman commented on sklearn/utils/_pprint.py:
```diff
@@ -0,0 +1,303 @@
+            indent += len(object.__class__.__name__)
+
+        if self._changed_only:
+            params = _changed_params(object)
```
not tested

jnothman commented on sklearn/utils/_pprint.py:
```diff
@@ -0,0 +1,303 @@
+        respect compact=True and will use one line per key-value if all cannot
+        fit in a single line.
+        Dict items will be rendered as <'key': value> while params will be
+        rendered as <key=value>. The implementation is mostly copy/pasting from
```
I'm not immediately convinced that we have sufficient attribution/licence here. Convince me

NicolasHug replied:
Hmm good point.

The [PSF License](https://en.wikipedia.org/wiki/Python_Software_Foundation_License) is BSD-like and non-copyleft, so I assume we can use the code here?

I'm really not expert on licenses so I might be wrong.

jnothman replied:
Most open source licences that are not copyleft merely require attribution and inclusion of the licence (what I'm not sure is whether linking to or referencing the licence is sufficient in some circumstances)

jnothman commented on sklearn/utils/_pprint.py:
```diff
@@ -0,0 +1,303 @@
+    r = getattr(typ, "__repr__", None)
+    if issubclass(typ, dict) and r is dict.__repr__:
+        if not object:
+            return "{}", True, False
```
not tested. should we aim for high coverage, or assume it will never be changed and hence not need to worry about regressions?

NicolasHug replied:
This part (and the others that I haven't fixed already) directly come from the built-in CPython code. In particular the parts related to the `context` variable deal with recursively nested objects like

```python
l = [1, 2]
l.append(l)
```

I can add tests for those to increase coverage, but I doubt we would ever hit such cases.

jnothman commented on sklearn/utils/_pprint.py:
```diff
@@ -0,0 +1,303 @@
+            return "{}", True, False
+        objid = id(object)
+        if maxlevels and level >= maxlevels:
+            return "{...}", False, objid in context
```
not tested

jnothman commented on sklearn/utils/_pprint.py:
```diff
@@ -0,0 +1,303 @@
+        if maxlevels and level >= maxlevels:
+            return "{...}", False, objid in context
+        if objid in context:
+            return pprint._recursion(object), False, True
```
not tested

jnothman commented on sklearn/utils/_pprint.py:
```diff
@@ -0,0 +1,303 @@
+
+
+def _safe_repr(object, context, maxlevels, level, changed_only=False):
+    """Same as the builtin _safe_repr, with added support for Estimator
```
needs attribution

NicolasHug committed:
Added test for changed_only and long lines

NicolasHug committed:
typo

NicolasHug committed:
Added authors names

NicolasHug commented:
Regarding attribution and license:

I added the original pprint module author's name to the file. Should I also include the whole CPython [License](https://github.com/python/cpython/blob/master/LICENSE) file  to `sklearn/utils/`? Or just part of it?

I don't know what would make sense.

jnothman commented:
> Should I also include the whole CPython License file to sklearn/utils/? Or just part of it?

It might be simplest, and most certainly legal, to put the full licence in that file. The licence also requests that you summarise any changes to the code.

NicolasHug committed:
Added license file

jnothman commented:
I've not yet looked through all the recent comments yet, but ipython truncates based on the number of elements, not the number of characters.

NicolasHug committed:
Added ellipsis based on number of elements in sequence + added increasinly aggressive repr strategies

NicolasHug committed:
Updated whatsnew

jnothman reviewed (commented):
Thanks for the good work @NicolasHug.

I think the main point of ellipsis is that the number of chars in the repr should be sub-linear with respect to the maximum length of any collection in the object, or more specifically that after a certain point of growth of a specific collection, repr size should be constant. I don't think the goal should necessarily be to fit within a fixed length (but if so, it should be a fixed line width and number of lines, rather than total number of characters), and I think we should make the behaviour pretty and useful, but not magical.

The basic test for ellipsis might be something like:

```py
pytest.mark.parametrize('generate', [
   lambda n: ['Hello'] * n,
   lambda n: set(range(n)),
   lambda n: {i: i for i in range(n)},
   lambda n: np.arange(n),
   lambda n: np.arange(n).reshape(-1, 1),
   lambda n: [{i: i for i in range(n)}],  # nested objects
])
def test(generate):
    assert len(repr(Estimator(generate(500)))) > len(repr(Estimator(generate(10))))
    assert repr(Estimator(generate(500))) == repr(Estimator(generate(1000)))
    assert ', ...' in repr(Estimator(generate(500)))
```

amueller commented:
To me the goal is to provide the most helpful and human-parsable string representations possible, and I suggested the current behavior.

Our current implementation has a character limit, and I think it makes some sense, since characters are a reasonable measure of complexity for parsing it.

So my goal is to provide as much useful information as possible with a limited amount of characters. The list of heuristics tries to achieve exactly that.

I might be biased because I really don't like the current repr and think it's basically useless (most people I talked to have developed a cognitive block to the output of ``fit`` in jupyter notebooks because the repr is not human-parsable).

If you don't like the idea of a character limit, we can also remove the heuristics.
I think some level of character limit is useful because it'll prevent some recursively defined estimator from filling some server hard drive with it's repr.

jnothman commented:
I'm not sure I appreciate the relevance of a character limit, but it's okay.

Nicolas if you don't believe my claim about having a constant limit with
the growth of a collection, I've not communicated it sufficiently, but
can't now.

I don't like automatic switching between showing all parameters and only
those that were changed. There is a user setting and it should be heeded.

I believe a simpler version should be road tested before a fancy strategy
for cutting down the repr.

NicolasHug committed:
dont use increaingly aggressive strategy

NicolasHug commented:
@jnothman , @amueller ,

I've updated `__repr__` to do the following:

- always apply ellipsis on sequences with a lot of elements
- if repr is too long, use bruteforce ellipsis (just like before)
- `changed_only` param is controlled by the env variable only.

amueller commented:
If you merge master the tests should pass (2.7 was removed from CI)

NicolasHug committed:
Merge branch 'master' into pr/9099

NicolasHug committed:
Fixed tests

NicolasHug committed:
Removed LICENSE file and put license text in _pprint.py

NicolasHug commented:
Merged master, hope I fixed the tests.

One thing we need to be aware of with the tests here, is that any change to the estimators used in `test_pprint.py` may break them (changing a default, adding / removing a parameter, etc.)

Broken tests are very easy to fix, but may occur often enough to be a recurring inconvenience.

NicolasHug committed:
fixed test_base

NicolasHug committed:
Sorted parameters dictionary for consistent output in 3.5

NicolasHug committed:
Actually using OrderedDict...

jnothman commented on sklearn/utils/_pprint.py:
```diff
@@ -0,0 +1,422 @@
+    init_func = getattr(estimator.__init__, 'deprecated_original',
+                        estimator.__init__)
+    init_params = signature(init_func).parameters
+    init_params = {name: param.default for (name, param) in
```
remove parentheses around `name, param`

jnothman commented on sklearn/utils/_pprint.py:
```diff
@@ -0,0 +1,422 @@
+from .._config import get_config
+
+
+# Dummy classes for correctly rendering key-value tuples from dicts or
```
put comment inside class?

jnothman commented on sklearn/utils/_pprint.py:
```diff
@@ -0,0 +1,422 @@
+    init_params = {name: param.default for (name, param) in
+                   init_params.items()}
+    for k, v in params.items():
+        if v != init_params[k]:
```
This won't work for `np.NaN`. We should probably handle that case. To handle instances that have been unpickled, we may need to directly test with `is_scalar_nan`. I.e. `if v != init_params[k] and not (is_scalar_nan(init_params[k]) and is_scalar_nan(v))`.

NicolasHug replied:
Good catch, thanks. I've added a test for this case

jnothman commented on sklearn/utils/_pprint.py:
```diff
@@ -0,0 +1,422 @@
+# reproduce, analyze, test, perform and/or display publicly, prepare
+# derivative works, distribute, and otherwise use Python alone or in any
+# derivative version, provided, however, that PSF's License Agreement and
+# PSF's notice of copyright, i.e., "Copyright (c) 2001, 2002, 2003, 2004,
```
I think this notice of copyright must be separately stated

jnothman commented on sklearn/utils/_pprint.py:
```diff
@@ -0,0 +1,422 @@
+# 3. In the event Licensee prepares a derivative work that is based on or
+# incorporates Python or any part thereof, and wants to make the derivative
+# work available to others as provided herein, then Licensee hereby agrees to
+# include in any such work a brief summary of the changes made to Python.
```
You haven't included a brief summary of changes to the original code.

jnothman commented on sklearn/utils/_pprint.py:
```diff
@@ -0,0 +1,422 @@
+        Dict items will be rendered as <'key': value> while params will be
+        rendered as <key=value>. The implementation is mostly copy/pasting from
+        the builtin _format_items().
+        This also adds ellipsis if the number of items is greated than
```
*greater

jnothman commented on sklearn/utils/tests/test_pprint.py:
```diff
@@ -0,0 +1,310 @@
+
+
+def test_gridsearch():
+    # render a gridsearch
```
Why is this of particular interest?

NicolasHug replied:
Just for the sake of illustrating what rendering a grid search looks like. Can remove if needed.

jnothman reviewed (commented):
This is very good, IMO!

NicolasHug committed:
Addressed comments

NicolasHug committed:
Added test for NaN changed parameter

jnothman reviewed (commented):
I've realised the change is a bit weak on documentation. Not sure where to put it in the user guide, though. Or whether `print_changed_only` should be illustrated in the example gallery.

Otherwise, I'm keen to see this road tested.

NicolasHug committed:
Update whatsnew

NicolasHug committed:
Added example to set_config()

NicolasHug commented on doc/whats_new/v0.21.rst:
```diff
@@ -185,12 +185,19 @@ Support for Python 3.4 and below has been officially dropped.
   ``max_depth`` were both specified by the user. Please note that this also
   affects all ensemble methods using decision trees.
-  :pr:`12344` by :user:`Adrin Jalali <adrinjalali>`.
+  :issue:`12344` by :user:`Adrin Jalali <adrinjalali>`.
```
@adrinjalali I took the liberty to fix this

NicolasHug commented:
I've added an example in `set_config()`. There's one in `config_context()` about `assume_finite` so I figured why not... But it's not necessarily ideal.

NicolasHug committed:
Removed example

jnothman commented:
We don't need another example of how to use config, we need somewhere in
the user guide that explicitly describes what this feature can do for the
user. An example in the gallery might be appropriate.

NicolasHug commented:
Where do I reference this example from in the doc?

NicolasHug committed:
Added example in gallery

amueller commented:
link it from the set_config docs maybe? I don't think there's a single section on that in the user guide, though...

NicolasHug commented:
Isn't it automatically linked?

amueller commented:
to the API docs, yes.

NicolasHug commented:
`set_config` is referenced in in `doc/modules/computing.rst` but I doubt it makes sense there

jnothman committed:
Spelling

jnothman reviewed (approved):
User guide could consider adding a section on working with scikit-learn in a terminal? But not in this PR.

jnothman commented:
Note that the example is automatically referenced at the bottom of https://41116-843222-gh.circle-artifacts.com/0/doc/modules/generated/sklearn.set_config.html

jnothman commented:
Yay! Thanks and well done @NicolasHug!!

GaelVaroquaux commented:
Hurray! Thanks heaps!! Sorry for not giving feedback on this. Anyhow, I love it.

Any plans to add back the color work, that @amueller had originally implemented? If it could be enabled by default, it would help usability.

amueller commented:
@GaelVaroquaux I'd love to see that, but I figured "small steps" would be better. This is already a big improvement imho.
I guess one question is whether we want something jupyter specific or something that also works on the console. I would probably go with jupyter and use html.

GaelVaroquaux commented:
> @GaelVaroquaux I'd love to see that, but I figured "small steps" would be
> better. This is already a big improvement imho.

+1!
